### ISSUE TITLE
wp.Struct zeros array grads when returned by wp.func

### ISSUE THREAD

**Issue by etaoxing** [NVIDIA/warp#1598]
### Bug Description

any wp.struct returned from a @wp.func zeroes reverse-mode gradients (A=no-struct works [8,16,24]; B and C both [0,0,0])

```python
"""Minimal repro: does Warp reverse-mode propagate gradients THROUGH a @wp.func that returns a wp.struct?

Three variants of the same computation  out = dot(s.v, s.v)  where s.v = 2 * x :
  A) no struct           -- baseline, grad should be d/dx dot(2x,2x) = 8x
  B) struct, scalars only -- struct{ v: vec3 }
  C) struct WITH a wp.array member (like mujoco_warp's collision_core.Geom)

If B and/or C return grad 0 while A is correct, that's the bug.
"""
import warp as wp


@wp.struct
class VecOnly:
  v: wp.vec3


@wp.struct
class VecAndArray:
  v: wp.vec3
  arr: wp.array(dtype=wp.float32)  # an (unused) array member -- mirrors Geom.vert/graph/...


@wp.func
def make_veconly(x: wp.vec3) -> VecOnly:
  s = VecOnly()
  s.v = 2.0 * x
  return s


@wp.func
def make_vecandarray(x: wp.vec3, dummy: wp.array(dtype=wp.float32)) -> VecAndArray:
  s = VecAndArray()
  s.v = 2.0 * x
  s.arr = dummy
  return s


@wp.kernel
def k_nostruct(x: wp.array(dtype=wp.vec3), out: wp.array(dtype=wp.float32)):
  v = 2.0 * x[0]
  out[0] = wp.dot(v, v)


@wp.kernel
def k_veconly(x: wp.array(dtype=wp.vec3), out: wp.array(dtype=wp.float32)):
  s = make_veconly(x[0])
  out[0] = wp.dot(s.v, s.v)


@wp.kernel
def k_vecandarray(x: wp.array(dtype=wp.vec3), dummy: wp.array(dtype=wp.float32), out: wp.array(dtype=wp.float32)):
  s = make_vecandarray(x[0], dummy)
  out[0] = wp.dot(s.v, s.v)


def run(name, kernel, extra_inputs):
  x = wp.array([wp.vec3(1.0, 2.0, 3.0)], dtype=wp.vec3, requires_grad=True)
  out = wp.zeros(1, dtype=wp.float32, requires_grad=True)
  tape = wp.Tape()
  with tape:
    wp.launch(kernel, dim=1, inputs=[x] + extra_inputs + [out])
  tape.backward(loss=out)
  g = x.grad.numpy()[0]
  expected = 8.0 * wp.vec3(1.0, 2.0, 3.0)  # d/dx dot(2x,2x) = 8x
  print(f"{name:28s} grad={g}  expected=[8 16 24]  {'OK' if abs(g[0] - 8.0) < 1e-5 else 'ZERO/WRONG <-- BUG'}")


if __name__ == "__main__":
  wp.init()
  print(f"warp {wp.config.version}")
  dummy = wp.zeros(1, dtype=wp.float32)
  run("A) no struct", k_nostruct, [])
  run("B) struct (vec3 only)", k_veconly, [])
  run("C) struct (vec3 + array)", k_vecandarray, [dummy])

```

### System Information

```bash
Software
  Warp (package):   1.14.0
  Warp (core):      1.14.0
  warp-clang:       1.14.0 (LLVM 15.0.7)
  NumPy:            2.2.6
  PyTorch:          2.12.1
  Python:           3.10.20 (main, Jun  2 2026, 22:39:36) [Clang 22.1.3 ]
  Platform:         macOS-26.5.1-arm64-arm-64bit

CUDA
  Enabled:          False
  Forward compat:   False

Libraries
  MathDx:           not available
  cuBQL:            enabled
  NanoVDB:          32.8.0

Build
  Compiler:         Clang 13.1.6
  Debug:            False
  Verify FP:        False
  Fast math:        False

CPU
  Model:            apple-m1
  ISA features:     0

Devices
  cpu
    Name:             arm
```

**Comment by daedalus5** [comment]
This was already fixed in commit [4025833](https://github.com/Mu-L/warp-gpu/commit/40258330a09504755f6e7b6dc187caa8216ea392)

Additional tests to prevent regression will be added, but closing this issue as it stands.

**Comment by shi-eric** [comment]
Progress update: f06dad0a1be3cac9e7dfd9cdccdb866fb3b8d984 landed regression coverage for reverse-mode gradients through a `wp.struct` returned by `@wp.func`.

The underlying behavior fix had already been noted when this issue was closed. This test-only follow-up now guards both reported struct shapes.

Public API behavior:

- No API was added or changed. Existing reverse-mode autodiff behavior is now covered when a kernel reads a `vec3` field from a struct returned by a kernel-callable `@wp.func`, including when the struct also contains an array member.

Test coverage:

- Added `warp/tests/test_struct.py::test_struct_returned_from_func_grad`.
- The test covers a `vec3`-only struct and a struct containing a `vec3` plus an unused array field.
- Both cases assert that `dot(2x, 2x)` produces `8x`, or `[8, 16, 24]` for the reported input.

The issue can remain closed.
